[MPS] Add Python Module Bindings for the MPS backend

razarmehr opened the pull request:
- This PR is a prerequisite for the upcoming Memory Leak Detection PR.
- Enable global manual seeding via `torch.manual_seed()` + test case
- Add `torch.mps.synchronize()` to wait for MPS stream to finish + test case
- Enable the following python interfaces for MPS:
  `torch.mps.[get_rng_state(), set_rng_state(), synchronize(), manual_seed(), seed()]`
- Added some test cases in test_mps.py
- Added `mps.rst` to document the `torch.mps` module.
- Fixed the failure with `test_public_bindings.py`

Description of new files added:
- `torch/csrc/mps/Module.cpp`: implements `torch._C` module functions for `torch.mps` and `torch.backends.mps`.
- `torch/mps/__init__.py`: implements Python bindings for `torch.mps` module.

razarmehr committed:
Add Python Module interface for MPS backend (#251)

- Enable global manual seeding via torch.manual_seed() + test case
- Add torch.mps.synchronize() to wait for MPS stream to finish + test case
- Enable the following python interfaces for MPS:
torch.mps.get_rng_state()
torch.mps.set_rng_state()
torch.mps.is_available()
torch.mps.synchronize()
torch.mps.manual_seed()
torch.mps.seed()
torch.mps.is_initialized()
torch.mps.init()

razarmehr committed:
Fix is_mps_available() regression on non-MPS devices (#268)

- This patch fixes a regression caused by recent MPS module interface. Since we now compile torch._C.is_mps_available()
only if USE_MPS is defined, then it may cause failures on CUDA (and other devices when USE_MPS is not defined) if we upstream.
So, this patch checks if is_mps_available is implemented first and then calls it.
- Also use the unique name `default_mps_generator` to avoid conflicts with CPU default generator

razarmehr committed:
Check of is_macos13_or_newer is built into torch.C_ bindings

razarmehr committed:
Fix lint issues

albanD commented on torch/mps/__init__.py:
```diff
@@ -0,0 +1,88 @@
+r"""
+This unit enables an interface for accessing MPS backend in python
```
Any reason for this new submodule instead of using the torch.backends.mps one?

razarmehr replied:
Yes, we want parity with CUDA and other modules. We should be consistent with naming the submodule namespaces.
Example, when CUDA synchronizes like `torch.cuda.synchronize()`, the user must be able to intuitively just replace `cuda` with `mps` to get the same operation with `torch.mps.synchronize()`.

albanD commented on torch/mps/__init__.py:
```diff
@@ -0,0 +1,88 @@
+    """
+    _lazy_init()
+
+def _lazy_init():
```
Why are you doing lazy initialization for this?

razarmehr replied:
The `lazy_init` is meant to initialize only "on-demand". The code is directly taken from CUDA's __init__.py:
https://github.com/pytorch/pytorch/blob/master/torch/cuda/__init__.py#L206

albanD replied:
But that is because cuda initialization is actually really expensive AND we don't initialize anything in cuda at import time (because cuda init allocates a couple GB in RAM and a couple 100s MB on the GPU).
My understanding here is that the only part that is lazy here is the MPS generator creation so:
- What happens if I just call `torch.rand(10, device="mps")`? This won't trigger this lazy init but actually use the mps generator?
- If initializing this generator is cheap (which I assume it is), I don't think it's worth the complexity.

razarmehr replied:
Updated the PR.

razarmehr commented on aten/src/ATen/detail/MPSHooksInterface.h:
```diff
@@ -28,13 +28,21 @@ struct TORCH_API MPSHooksInterface {
   }
 
+  virtual bool isOnMacOS13orNewer() const {
+    return false;
```
It's meant to be used with MPSHooks only. If MPS isn't available, running this should return false always:
`at::detail::getMPSHooks().isOnMacOS13orNewer()`.
It's a way to know if MPS supports the APIs that were added from MacOS 13.

albanD replied:
Sure but why should you always assume you're on macos older than MacOS 13 if MPS is not available? The two sound quite unrelated. So why no always return true?

My point is that given that this is an arbitrary choice, why do we pick one instead of raising an error like we do for the other methods below.

malfet replied:
I think we should follow the pattern where all functions other than `hasMPS()` should raise an assert, as they should never be called if MPS is not available.
```suggestion
    AT_ERROR("MPS backend is not enabled");
```

razarmehr replied:
Updated the PR.

razarmehr commented:
I'm working on a fix to address CI failures on other backends caused by this PR.

razarmehr commented on torch/csrc/api/include/torch/mps.h:
```diff
@@ -0,0 +1,21 @@
+#include <cstdint>
+
+namespace torch {
+namespace mps {
```
This is supposed to be the C++ binding and interface (not python), so it has nothing to do with the Module.cpp.
These C++ APIs are meant to be the equivalent of CUDA ones (in the same paths) to be accessible for C++ users:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/include/torch/cuda.h
https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/src/cuda.cpp

razarmehr replied:
Removed the mps.h and mps.cpp.

razarmehr commented on torch/csrc/mps/Module.cpp:
```diff
@@ -0,0 +1,85 @@
+
+static PyObject* MPSModule_initExtension(PyObject* self, PyObject* noargs) {
+#if C10_ASAN_ENABLED
+  TORCH_WARN(
```
> This needs to be inside the HANDLE_TH_ERRORS macro otherwise it won't be translated to a python warning properly.

Code is taken from CUDA. The `HANDLE_TH_ERRORS` is at line 21 below.

> Also this is going to be triggered as soon as the user import torch

This is meant to be run on-demand only from `lazy_init()` via `torch._C._mps_init()`. By on-demand, I mean only when user for example calls `torch.mps.manual_seed()`, then if the module isn't initialized already, then it'll initialize it lazily.

albanD replied:
> Code is taken from CUDA. The HANDLE_TH_ERRORS is at line 21 below.

Well the CUDA code is wrong and should be changed :(

razarmehr replied:
Removed all the redundant init code.

malfet commented on torch/mps/__init__.py:
```diff
@@ -0,0 +1,88 @@
+        torch._C._mps_init()
+        _initialized = True
+
+@lru_cache()
+def is_available() -> bool:
+    r"""Returns a bool indicating if MPS is currently available."""
+    if not hasattr(torch._C, '_is_mps_available'):
+        return False
+    return torch._C._is_mps_available()
```
Why implement it again, rather than call the function from backends?
```suggestion
def is_available() -> bool:
    r"""Returns a bool indicating if MPS is currently available."""
    return torch.backends.mps.is_available()
```

razarmehr replied:
Removed is_available() and is_macos13_or_newer from torch.mps, as they already exist in torch.backends.mps.

malfet commented on torch/backends/mps/__init__.py:
```diff
@@ -15,12 +15,16 @@ def is_built() -> bool:
 @_lru_cache()
 def is_available() -> bool:
     r"""Returns a bool indicating if MPS is currently available."""
+    if not hasattr(torch._C, '_is_mps_available'):
+        return False
```
Shouldn't be needed, as it works fine right now, see:
```
$ python -c "import torch;import sys;print(sys.platform, torch.__version__, torch.backends.mps.is_available())"
linux 2.0.0a0+git208f330 False

```
```suggestion
```

razarmehr replied:
Updated the PR.

malfet commented on torch/backends/mps/__init__.py:
```diff
@@ -15,12 +15,16 @@ def is_built() -> bool:
 @_lru_cache()
 def is_macos13_or_newer() -> bool:
     r"""Returns a bool indicating whether MPS is running on MacOS 13 or newer."""
+    if not hasattr(torch._C, '_is_mps_on_macos_13_or_newer'):
+        return False
```
Again, shouldn't be needed, though I wonder if this call should really trigger an exception on non-MPS capable systems:
```
$ python -c "import torch;import sys;print(sys.platform, torch.__version__, torch.backends.mps.is_macos13_or_newer())"
linux 2.0.0a0+git208f330 False
```
```suggestion
```

razarmehr replied:
Updated the PR.

malfet commented on torch/mps/__init__.py:
```diff
@@ -0,0 +1,88 @@
+r"""
+This unit enables an interface for accessing MPS backend in python
```
```suggestion
This package enables an interface for accessing MPS backend in python
```

malfet commented on torch/mps/__init__.py:
```diff
@@ -0,0 +1,88 @@
+        return False
+    return torch._C._is_mps_available()
+
+@lru_cache()
+def is_macos13_or_newer() -> bool:
+    r"""Returns a bool indicating whether MPS is running on MacOS 13 or newer."""
+    if not hasattr(torch._C, '_is_mps_on_macos_13_or_newer'):
+        return False
+    return torch._C._is_mps_on_macos_13_or_newer()
```
Why this needs to be part of `torch.mps` package? Isn't it an implementation detail that usually specific to a backend?

razarmehr replied:
Updated the PR.

malfet commented on torch/csrc/mps/Module.cpp:
```diff
@@ -0,0 +1,85 @@
+    }
+  };
+
+  // NOLINTNEXTLINE(performance-unnecessary-copy-initialization)
+  auto gen = at::mps::detail::getDefaultMPSGenerator();
+  auto default_mps_generator =
+      (THPGenerator*)THPGenerator_initDefaultGenerator(gen);
+  set_module_attr("default_mps_generator", (PyObject*)default_mps_generator);
+
```
This feels like an anti-pattern: we should not litter public namespace with native objects, but rather make a function call (say `torch._C.get_mps_default_generator`) and call it from appropriate functions in python

razarmehr replied:
Updated the PR.

malfet commented on torch/csrc/mps/Module.cpp:
```diff
@@ -0,0 +1,85 @@
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject* MPSModule_isAvailable(PyObject* _unused, PyObject* noargs) {
+  HANDLE_TH_ERRORS
+  if (at::detail::getMPSHooks().hasMPS()) {
+    Py_RETURN_TRUE;
+  } else {
+    Py_RETURN_FALSE;
+  }
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject* MPSModule_isMacOS13orNewer(PyObject* _unused, PyObject* noargs) {
+  HANDLE_TH_ERRORS
+  if (at::detail::getMPSHooks().isOnMacOS13orNewer()) {
+    Py_RETURN_TRUE;
+  } else {
+    Py_RETURN_FALSE;
+  }
+  END_HANDLE_TH_ERRORS
+}
```
I've already exposed those hooks in the `torch.backends.mps`, why exposing them again?

razarmehr replied:
I'll remove these ones.

razarmehr replied:
Updated the PR.

malfet reviewed (changes requested):
Some parts of the code is re-exporting MPS properties which are already available in `torch.backends.mps` package and also copy-n-pastes anti-patterns from old days of CUDA.(like lazy-init)
 Please request re-review when this feedback is addressed.

razarmehr committed:
Check for MPS availability before adding generator

razarmehr committed:
Removed lazy initialization
- Removed USE_MPS condition when adding python binding definitions
- Replaced the global attribute setting of default_mps_generator with torch.C_._mps_get_default_generator()
- Removed redundant hasattr() checks
- Error out in MPSHooks::isOnMacOS13orNewer() if MPS not available
- Removed is_available() and is_macos13_or_newer() from torch.mps (already exists in torch.backends.mps)

razarmehr committed:
Fix lint errors

razarmehr commented on torch/backends/mps/__init__.py:
```diff
@@ -15,12 +15,16 @@ def is_built() -> bool:
 @_lru_cache()
 def is_available() -> bool:
     r"""Returns a bool indicating if MPS is currently available."""
+    if not hasattr(torch._C, '_is_mps_available'):
```
Updated the PR.

razarmehr commented on torch/csrc/Module.cpp:
```diff
@@ -1274,6 +1274,9 @@ PyObject* initModule() {
 #ifdef USE_CUDA
   THPUtils_addPyMethodDefs(methods, THCPModule_methods());
 #endif
+#ifdef USE_MPS
```
Updated the PR.

razarmehr committed:
Remove the C++ APIs mps.cpp and mps.h

razarmehr committed:
Add terminating nullptr at the end of _MPSModule_methods list

razarmehr committed:
Add mps.rst for docs

razarmehr committed:
Asd orphan to mps.rst

razarmehr committed:
Add mps to index.rst

razarmehr committed:
Fix document issue for manual_seed()

razarmehr committed:
Add underscore to make _get_default_mps_generator() private()
This should fix the failure with test_public_bindings.py

albanD commented on torch/csrc/Module.cpp:
```diff
@@ -1219,6 +1215,8 @@ void initIttBindings(PyObject* module);
 } // namespace torch
 #endif
 
+PyMethodDef* MPSModule_methods();
```
If you're not using a header for this, leave a comment where it will be added.

razarmehr replied:
I think a new header for the single line `PyMethodDef* MPSModule_methods();`  is overkill. 
So do you want me to put a TODO comment like this?
```
// TODO: add module method declarations to a common header file
PyMethodDef* MPSModule_methods();
```

razarmehr replied:
Moved it to a header, similar to what we do with other modules like [csrc/multiprocessing/init.h](https://github.com/pytorch/pytorch/blob/master/torch/csrc/multiprocessing/init.h)

albanD commented:
Sounds pretty good. Thanks for the changes. Only small things.

razarmehr commented on test/test_mps.py:
```diff
@@ -25,6 +25,7 @@
 from torch.testing import make_tensor
 from torch.testing._comparison import TensorLikePair
 from torch.testing._internal.common_dtype import get_all_dtypes, integral_types
+import torch.mps
```
It's used in `test_default_mps_generator()` and `test_device_synchronize()`, to test` torch.mps.manual_seed()`, `torch.mps.synchrnonize()` and etc.
I can add a comment in front of this import.

razarmehr replied:
Removed it.

razarmehr committed:
Add new header for python_functions() of MPS module
- Remove import torch.mps from test_mps
- Add torch::mps namespace

albanD reviewed (approved):
Sounds good!
Thanks for the updates.

razarmehr commented:
@pytorchbot rebase

razarmehr commented:
The dynamo CI failure is unrelated to this PR and is due to following failed issue #89395.
```
FAIL [0.001s]: test_coalesce_reference_cycle_cpu_float64 (__main__.TestSparseCPU)
```

razarmehr commented:
@pytorchbot merge -f "Lint and MPS tests are green"

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your PR, but it seems to break MacOS test in trunk https://hud.pytorch.org/pytorch/pytorch/commit/bae397ec63aefef60fa9e0967ef28ecf7b954e34' -c nosignal

huydhn commented:
The failed test can be reproduce by running `pytest -k test_fd_limit_exceeded test_dataloader.py`.  Here is the log FYI:

```
=================================================================== test session starts ====================================================================
platform darwin -- Python 3.9.15, pytest-7.2.0, pluggy-1.0.0
rootdir: /Users/huydo/Storage/mine/pytorch, configfile: pytest.ini
plugins: xdoctest-1.1.0, shard-0.1.2, flakefinder-1.1.0, hypothesis-6.56.4, rerunfailures-10.2, xdist-3.0.2, repeat-0.9.1
collecting 0 items                                                                                                                                         /Users/huydo/Storage/mine/pytorch/test/test_dataloader.py:681: PytestCollectionWarning: cannot collect test class 'TestProperExitDataset' because it has a __init__ constructor (from: test/test_dataloader.py)
  class TestProperExitDataset(Dataset):
/Users/huydo/Storage/mine/pytorch/test/test_dataloader.py:698: PytestCollectionWarning: cannot collect test class 'TestProperExitIterableDataset' because it has a __init__ constructor (from: test/test_dataloader.py)
  class TestProperExitIterableDataset(IterableDataset):
/Users/huydo/Storage/mine/pytorch/test/test_dataloader.py:801: PytestCollectionWarning: cannot collect test class 'TestWorkerInfoDataset' because it has a __init__ constructor (from: test/test_dataloader.py)
  class TestWorkerInfoDataset(SynchronizedDataset):
/Users/huydo/Storage/mine/pytorch/test/test_dataloader.py:902: PytestCollectionWarning: cannot collect test class 'TestMultiEpochDataset' because it has a __init__ constructor (from: test/test_dataloader.py)
  class TestMultiEpochDataset(IterableDataset):
/Users/huydo/Storage/mine/pytorch/test/test_dataloader.py:2613: PytestCollectionWarning: cannot collect test class 'TestWorkerQueueDataset' because it has a __init__ constructor (from: test/test_dataloader.py)
  class TestWorkerQueueDataset(Dataset):
collected 164 items / 162 deselected / 2 selected
Running 2 items in this shard

test_dataloader.py objc[81968]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called.
objc[81968]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
Traceback (most recent call last):
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 1134, in _try_get_data
    data = self._data_queue.get(timeout=timeout)
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/multiprocessing/queues.py", line 113, in get
    if not self._poll(timeout):
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/multiprocessing/connection.py", line 262, in poll
    return self._poll(timeout)
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/multiprocessing/connection.py", line 429, in _poll
    r = wait([self], timeout)
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/multiprocessing/connection.py", line 936, in wait
    ready = selector.select(timeout)
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/selectors.py", line 416, in select
    fd_event_list = self._selector.poll(timeout)
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/site-packages/torch/utils/data/_utils/signal_handling.py", line 66, in handler
    _error_if_any_worker_fails()
RuntimeError: DataLoader worker (pid 81968) is killed by signal: Abort trap: 6.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<string>", line 23, in <module>
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 635, in __next__
    data = self._next_data()
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 1330, in _next_data
    idx, data = self._get_data()
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 1296, in _get_data
    success, data = self._try_get_data()
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 1147, in _try_get_data
    raise RuntimeError('DataLoader worker (pid(s) {}) exited unexpectedly'.format(pids_str)) from e
RuntimeError: DataLoader worker (pid(s) 81968) exited unexpectedly

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 28, in <module>
AssertionError
Fobjc[81986]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called.
objc[81986]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
Traceback (most recent call last):
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 1134, in _try_get_data
    data = self._data_queue.get(timeout=timeout)
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/multiprocessing/queues.py", line 113, in get
    if not self._poll(timeout):
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/multiprocessing/connection.py", line 262, in poll
    return self._poll(timeout)
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/multiprocessing/connection.py", line 429, in _poll
    r = wait([self], timeout)
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/multiprocessing/connection.py", line 936, in wait
    ready = selector.select(timeout)
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/selectors.py", line 416, in select
    fd_event_list = self._selector.poll(timeout)
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/site-packages/torch/utils/data/_utils/signal_handling.py", line 66, in handler
    _error_if_any_worker_fails()
RuntimeError: DataLoader worker (pid 81986) is killed by signal: Abort trap: 6.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<string>", line 23, in <module>
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 635, in __next__
    data = self._next_data()
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 1330, in _next_data
    idx, data = self._get_data()
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 1296, in _get_data
    success, data = self._try_get_data()
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 1147, in _try_get_data
    raise RuntimeError('DataLoader worker (pid(s) {}) exited unexpectedly'.format(pids_str)) from e
RuntimeError: DataLoader worker (pid(s) 81986) exited unexpectedly

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 28, in <module>
AssertionError
F                                                                                                                                [100%]

========================================================================= FAILURES =========================================================================
__________________________________________________________ TestDataLoader.test_fd_limit_exceeded ___________________________________________________________
Traceback (most recent call last):
  File "/Users/huydo/Storage/mine/pytorch/test/test_dataloader.py", line 1039, in test_fd_limit_exceeded
    subprocess.check_output([sys.executable, '-c', """\
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/subprocess.py", line 424, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/Users/huydo/miniconda3/envs/py3.9/bin/python', '-c', 'import torch\nimport resource\nfrom torch.utils.data import DataLoader, IterableDataset\n\nclass RandomDataset(IterableDataset):\n    def __init__(self, len, size):\n        super(RandomDataset).__init__()\n        self.len = len\n        self.size = size\n\n    def __iter__(self):\n        return self\n\n    def __next__(self):\n        if self.len <= 0:\n            raise StopIteration\n        self.len -= 1\n        return torch.randn(self.size)\n\ntry:\n    keep_fds_alive = []\n    resource.setrlimit(resource.RLIMIT_NOFILE, (100, 100))\n    for random_t in DataLoader(RandomDataset(200, (2,2)), multiprocessing_context="fork",\n                               num_workers=1):\n      random_t.max(dim=0)\n      keep_fds_alive.append(random_t)\nexcept RuntimeError as e:\n    assert "ulimit -n" in str(e)\n    assert "set_sharing_strategy" in str(e)\n']' returned non-zero exit status 1.
__________________________________________________ TestDataLoaderPersistentWorkers.test_fd_limit_exceeded __________________________________________________
Traceback (most recent call last):
  File "/Users/huydo/Storage/mine/pytorch/test/test_dataloader.py", line 2411, in test_fd_limit_exceeded
    subprocess.check_output([sys.executable, '-c', """\
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/subprocess.py", line 424, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/Users/huydo/miniconda3/envs/py3.9/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/Users/huydo/miniconda3/envs/py3.9/bin/python', '-c', 'import torch\nimport resource\nfrom torch.utils.data import DataLoader, IterableDataset\n\nclass RandomDataset(IterableDataset):\n    def __init__(self, len, size):\n        super(RandomDataset).__init__()\n        self.len = len\n        self.size = size\n\n    def __iter__(self):\n        return self\n\n    def __next__(self):\n        if self.len <= 0:\n            raise StopIteration\n        self.len -= 1\n        return torch.randn(self.size)\n\ntry:\n    keep_fds_alive = []\n    resource.setrlimit(resource.RLIMIT_NOFILE, (100, 100))\n    for random_t in DataLoader(RandomDataset(200, (2,2)), multiprocessing_context="fork",\n                               num_workers=1, persistent_workers=True):\n      random_t.max(dim=0)\n      keep_fds_alive.append(random_t)\nexcept RuntimeError as e:\n    assert "ulimit -n" in str(e)\n    assert "set_sharing_strategy" in str(e)\n']' returned non-zero exit status 1.
================================================================= short test summary info ==================================================================
FAILED test_dataloader.py::TestDataLoader::test_fd_limit_exceeded - subprocess.CalledProcessError: Command '['/Users/huydo/miniconda3/envs/py3.9/bin/python', '-c', 'import torch\nimport resource\nfrom torch.utils.data i...
FAILED test_dataloader.py::TestDataLoaderPersistentWorkers::test_fd_limit_exceeded - subprocess.CalledProcessError: Command '['/Users/huydo/miniconda3/envs/py3.9/bin/python', '-c', 'import torch\nimport resource\nfrom torch.utils.data i...
============================================================ 2 failed, 162 deselected in 10.62s ============================================================
```

razarmehr committed:
Fix the failure with multiprocessing dataloader caused by bad forking

razarmehr commented:
@pytorchbot rebase

razarmehr commented:
Thanks @huydhn. The commit 2ba30c2 should fix the problem with the bad forking problem. I tested with `pytest -k test_fd_limit_exceeded test_dataloader.py`, and it's all good now.
Since that dataloader test isn't part of the CI checks on the PR, I will re-merge the PR with that fix (after all CIs are green), and will monitor PyTorch HUD for any regressions.

huydhn commented:
You can add `ciflow/trunk` and it will run all MacOS tests on trunk in the PR for you, including the failed test above.  It's also done automatically if the PR is merged normally with `@pytorchbot merge` instead of force merging

razarmehr commented:
@pytorchbot merge -g

kulinseth reviewed (commented):
Looks good.
